upstairs/src/volume.rs

```diff
@@ -0,0 +1,1228 @@
+
+    fn get_block_size(&self) -> Result<u64, CrucibleError> {
+        Ok(self.block_size)
+    }
+
+    fn get_uuid(&self) -> Result<Uuid, CrucibleError> {
+        Ok(self.uuid)
+    }
+
+    fn read(
+        &self,
+        offset: Block,
+        data: Buffer,
+    ) -> Result<BlockReqWaiter, CrucibleError> {
+        // In the cast that this volume only has a read only parent, serve
+        // reads directly from that
+        if self.sub_volumes.is_empty() {
+            if let Some(read_only_parent) = &self.read_only_parent {
+                return read_only_parent.read(offset, data.clone());
+            } else {
+                crucible_bail!(
+                    CannotServeBlocks,
+                    "No read only parent, no sub volumes!",
+                );
+            }
+        }
+
+        let affected_sub_volumes = self.sub_volumes_for_lba_range(
+            offset.value,
+            data.len() as u64 / self.block_size,
+        );
+
+        // TODO parallel dispatch!
+        let mut data_index = 0;
+        for (coverage, sub_volume) in affected_sub_volumes {
+            let sub_offset = Block::new(
+                sub_volume.compute_sub_volume_lba(coverage.start),
+                offset.shift,
+            );
+            // 0..10 would be size 10
+            let sz = (coverage.end - coverage.start) as usize
+                * self.block_size as usize;
+            let sub_buffer = Buffer::new(sz);
+
+            let mut waiter = sub_volume.read(sub_offset, sub_buffer.clone())?;
+            waiter.block_wait()?;
+
+            // When performing a read, check the parent coverage: if it's
+            // Some(range), then we have to perform multiple reads:
+            //
+            // - if the sub volume block is "owned", then return it
+            // - otherwise, return the read only parent block
+            let parent_coverage = self.read_only_parent_for_lba_range(
+                coverage.start,
+                coverage.end - coverage.start,
+            );
+
+            if let Some(parent_coverage) = parent_coverage {
+                let parent_offset = Block::new(coverage.start, offset.shift);
+                let sz = self.block_size as usize
+                    * (parent_coverage.end - parent_coverage.start) as usize;
+
+                let parent_buffer = Buffer::new(sz);
+                let mut waiter = self
+                    .read_only_parent
+                    .as_ref()
+                    .unwrap()
+                    .read(parent_offset, parent_buffer.clone())?;
+                waiter.block_wait()?;
+
+                let mut data_vec = data.as_vec();
+                let sub_data_vec = sub_buffer.as_vec();
+                let sub_owned_vec = sub_buffer.owned_vec();
+                let parent_data_vec = parent_buffer.as_vec();
+
+                // "ownership" comes back from the downstairs per byte but all
+                // writes occur per block. iterate over blocks here.
```

Comment:
I have questions about this part (this comment is more for me so I can find it)

Reply:
In `GtoS::transfer`:

```
                        for i in &response.data {                               
                            vec[offset] = *i;                                   
                            owned_vec[offset] = !response.hashes.is_empty();    
                            offset += 1;                                        
                        }
```

owned_vec is populated per byte but the value doesn't change for the whole response.

Reply:
I think we talked about this only needing to check at a block level, so maybe you already did that and if not we can iterate into it.